File: src/Nap.Immutable/Cookies/NapCookie.fs
namespace Nap

open System
open System.Collections.Generic
open System.Net
open System.Linq

/// <summary>
/// A single cookie returned from the server.
/// </summary>
type NapCookie =
    {
        /// <summary>
        /// Gets the name of the cookie.
        /// </summary>
        Name: string

        /// <summary>
        /// Gets the value of the cookie.
        /// </summary>
        Value: string

        /// <summary>
        /// Gets all metadata associated with this cookie.
        /// </summary>
        Metadata: NapCookieMetadata
    }
    with
    /// <summary>
    /// Initializes a new instance of a <see cref="NapCookie"/> class.
    /// </summary>
    /// <param name="requestUri">The request URI that this cookie belongs to.</param>
    /// <param name="cookieString">The string that will generate a cookie (value of a Set-Cookie header).</param>
    static member Create (requestUri:Uri) (cookieString: string) =
        let lowercaseSegmentKeys =
            Map.toSeq >> Seq.map (fun ((k:string),v) -> k.ToLower(),v) >> Map.ofSeq
        let segments =
            seq { for c in cookieString.Split(';') do
                     let split = c.Split('=') |> Seq.map(fun s -> s |> Option.ofObj |> Option.map (fun o -> o.Trim()) |> Option.toObj) |> Seq.toList
                     match split with
                     | [name] -> yield (name, None)
                     | name::rest -> yield (name, Some(rest.Head))
                     | [] -> ()
            }
        if (segments |> Seq.length > 0) then
            {
                Name = (segments |> Seq.head) |> fst
                Value = 
                    match (segments |> Seq.head) |> snd with
                    | Some(v) -> v
                    | None -> String.Empty
                Metadata = NapCookieMetadata.Create requestUri (lowercaseSegmentKeys <| Map.ofSeq segments)
            }
            else
            {
                Name = String.Empty
                Value = String.Empty
                Metadata = NapCookieMetadata.Create requestUri (lowercaseSegmentKeys <| Map.ofSeq segments)
            }

    static member FromCookie (cookie:Cookie) =
        {
            Name = cookie.Name
            Value = cookie.Value
            Metadata = { CreationDate = cookie.TimeStamp
                         Expires = cookie.Expires |> Some
                         MaxAge = None
                         Domain = cookie.Domain
                         Path = cookie.Path
                         IsSecure = cookie.Secure
                         HttpOnly = cookie.HttpOnly
                         SameSite = SameSitePolicy.Unset
                       }
        }

File: src/Nap/NapCookie.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;

namespace Nap
{
    /// <summary>
    /// A single cookie returned from the server.
    /// </summary>
    public struct NapCookie
    {
        /// <summary>
        /// Gets the name of the cookie.
        /// </summary>
        public string Name { get; }

        /// <summary>
        /// Gets the value of the cookie.
        /// </summary>
        public string Value { get; }

        /// <summary>
        /// Gets all metadata associated with this cookie.
        /// </summary>
        public NapCookieMetadata Metadata { get; }

        /// <summary>
        /// Initializes a new instance of a <see cref="NapCookie"/> class.
        /// </summary>
        /// <param name="requestUri">The request URI that this cookie belongs to.</param>
        /// <param name="cookieString">The string that will generate a cookie (value of a Set-Cookie header).</param>
        public NapCookie(Uri requestUri, string cookieString)
        {
            var segments = cookieString.Split(';').Select(c =>
            {
                var split = c.Split('=').Select(s => s?.Trim()).ToList();
                if (c.Length > 1)
                    return new KeyValuePair<string, string>(split[0], split[1]);
                else
                    return new KeyValuePair<string, string>(split[0], null);
            }).ToDictionary(kv => kv.Key, kv => kv.Value);
            if (segments.Count > 0)
            {
                Name = segments.First().Key;
                Value = segments.First().Value;
                Metadata = new NapCookieMetadata(requestUri, segments);
            }
            else
            {
                Name = null;
                Value = null;
                Metadata = new NapCookieMetadata();
            }
        }

        /// <summary>
        /// Initializes a new instance of a <see cref="NapCookie"/> class from a .NET cookie.
        /// </summary>
        public NapCookie(Cookie cookie)
        {
            Name = cookie.Name;
            Value = cookie.Value;
            Metadata = new NapCookieMetadata(cookie);
        }
    }
}
